Blog post: Unbearable lightness of refactoring post (#586)
artsy/artsy.github.io: _posts/2019-09-10-unbearable-lightness-of-refactoring.markdown

---
layout: epic
title: "Unbearable Lightness of Refactoring"
date: "2019-09-10"
author: [ashkan]
categories: [refactoring, community, culture]
comment_id: 522
---

> As engineers we are constantly in the process of building new features and improve our existing ones. Nowadays, with the help of tools and processes like code reviews one could argue the quality of the codes being written has risen. At Artsy a pull request normally has one Assignee and possibly one or more Reviewers, so why we still do lot of refactoring?

Reviewer: ```suggestion
As engineers we are constantly in the process of building new features and improving our existing ones. Nowadays, with the help of tools and processes like code reviews one could argue the quality of the code being written has risen. At Artsy a pull request normally has one Assignee and possibly one or more Reviewers, so why we still do lot of refactoring?
```

Reviewer: I think this section should be reworked. 

I see what you're getting at... even though we're continually writing code, we use tools like code reviews to ensure we're outputting high quality code. We still have to do a lot of refactoring. Why?

The issue, I think, is that it seems like the blog post will address this issue specifically... why do we need to do refactoring when we should be producing high quality code? 

Maybe instead you could write about how `quality worth of art` is one of our company principles and that refactoring is a good mechanism to achieve that.

> As engineers we are constantly in the process of building new features and improving our existing ones. Nowadays, with the help of tools and processes like code reviews one could argue the quality of the code being written has risen. At Artsy a pull request normally has one Assignee and possibly one or more Reviewers, so why we still do lot of refactoring?

Reviewer: I think this section should be reworked.

I see what you're getting at... even though we're continually writing code, we use tools like code reviews to ensure we're outputting high quality code. We still have to do a lot of refactoring. Why?

The issue, I think, is that it seems like the blog post will address this issue specifically... why do we need to do refactoring when we should be producing high quality code?

Maybe instead you could write about how quality worth of art is one of our company principles and that refactoring is a good mechanism to achieve that.

Author: good point, the idea behind the blogpost is to ask question why we still do constant refactoring if the quality of the code we are writing at a time matches our standards and is `Quality of worthy of Art` and also make some suggestions about how to do a good isolated refactoring.
The intro paragraph is basically raising the question and im trying to answer it in next few paragraphs but i agree this could be more structured.

> There is no means of testing which decision is better, because there is no basis for comparison. We live everything as it comes, without warning, like an actor going on cold. And what can life be worth if the first rehearsal for life is life itself?

― Milan Kundera, [The Unbearable Lightness of Being](https://en.wikipedia.org/wiki/The_Unbearable_Lightness_of_Being)

Part of me wants to end this blogpost by Kundra’s quote, but for now lets get deeper.

<!-- more -->

## "The Refactor"

> Recently we've started adding SCA (Strong Customer Authentication) support to one of our services. This service is relatively young in our stack and very well revieiwed. While the original code and approach looked nice and simple, as this service naturally grew and we started adding more logic to it, things got more and more complicated. During SCA support efforts, we realized its time to refactor. The code I was trying to refactor was less than a year old and originally written, well..., by me!

Reviewer 2: > originally written, well... by me!

Should I be worried? embarrassed? Well, not really. As engineers when we build things we tend to look at current state of affairs, we attempt to predict the future as much as possible but the future is always changing, moving. SCA feature we were about to add to our existing logic wasn't a requirement a year ago (a year ago I didn't know what SCA was). So first thing, never be embarrassed about refactoring, because the thing you're working on is often entirely unknown and you can't expect to get something unknown totally right the first time around. Looked at in this light refactoring is healthy.

Back to our SCA change, we ended up having some discussions about how to improve our already complicated logic in a way that easily supports future updates. We started by trying to separate the different steps that an order has to go through to fully get submitted and try to simplify it, weighing the pros and cons of each approach. While pretty much all of our solutions would have some cons, we managed to find our best option and started refactoring. We basically went from:

> ```ruby

Reviewer: It'd be nice to break this down a little bit. Isolate specific parts of the refactor and talk about why you did them.

---
layout: epic
title: "Unbearable Lightness of Refactoring"
date: "2019-09-10"
author: [ashkan]
categories: [refactoring, community, culture]
comment_id: 522
---

As engineers we are constantly in the process of building new features and improving our existing ones. Nowadays, with the help of tools and processes like code reviews one could argue the quality of the code being written has risen. At Artsy a pull request normally has one Assignee and possibly one or more Reviewers, so why we still do lot of refactoring?

> There is no means of testing which decision is better, because there is no basis for comparison. We live everything as it comes, without warning, like an actor going on cold. And what can life be worth if the first rehearsal for life is life itself?

> ― Milan Kundera, [The Unbearable Lightness of Being](https://en.wikipedia.org/wiki/The_Unbearable_Lightness_of_Being)

Reviewer 3: If you prefix the blank line and attribution, they'll get included in the blockquote:

```markdown
> ... life itself?
> 
> – Milan Kundera...
```

Part of me wants to end this blogpost by Kundra’s quote, but for now lets get deeper.

<!-- more -->

## "The Refactor"

> Recently we've started adding SCA (Strong Customer Authentication) support to one of our services. This service is relatively young in our stack and very well reviewed. While the original code and approach looked nice and simple, as this service naturally grew and we started adding more logic to it, things got more and more complicated. During SCA support efforts, we realized its time to refactor. The code I was trying to refactor was less than a year old and originally written, well... by me!

Reviewer 3: ```suggestion
Recently we've started adding [SCA (Strong Customer Authentication)](https://stripe.com/docs/strong-customer-authentication) support to one of our services. This service is relatively young in our stack and very well reviewed. While the original code and approach looked nice and simple, as this service naturally grew and we started adding more logic to it, things got more and more complicated. During SCA support efforts, we realized its time to refactor. The code I was trying to refactor was less than a year old and originally written, well... by me!
```

Reviewer 3: Up to you, but `originally written, well... by me!` could link to https://twitter.com/davidwalshblog/status/953663412013293569

> Should I be worried? embarrassed? Well, not really. As engineers when we build things we tend to look at current state of affairs, we attempt to predict the future as much as possible but the future is always changing, moving. SCA feature we were about to add to our existing logic wasn't a requirement a year ago (a year ago I didn't know what SCA was). So first thing, never be embarrassed about refactoring, because the thing you're working on is often entirely unknown and you can't expect to get something unknown totally right the first time around. Looked at in this light refactoring is healthy.

Reviewer 3: ```suggestion
Should I be worried? Embarrassed? Well, not really. As engineers, when we build things we tend to look at current state of affairs, and we attempt to predict the future as much as possible. But the future is always changing, moving. SCA feature we were about to add to our existing logic weren't a requirement a year ago (a year ago I didn't know what SCA was). So first thing, **never be embarrassed about refactoring**, because the thing you're working on is often entirely unknown and you can't expect to get something unknown totally right the first time around. Looked at it in this light, refactoring is healthy.
```

This is just some syntax/grammar stuff, and bolding your main point (which is a really great point).

### Lets Get More Specific

> Back to our SCA change, we ended up having some discussions about how to improve our already complicated logic in a way that easily supports future updates. We started by trying to separate the different steps that an order has to go through to fully get submitted and try to simplify it, weighing the pros and cons of each approach. While pretty much all of our solutions would have some cons, we managed to find our best option and started refactoring. Without getting too technical, lets look at the actual change.

Reviewer 3: ```suggestion
Back to our SCA change, we ended up having some discussions about how to improve our already complicated logic in a way that easily supports future updates. We started by trying to separate the different steps that an order has to go through to fully get submitted and try to simplify it, weighing the pros and cons of each approach. While pretty much all of our solutions would have some disadvantages, we managed to find our best option and started refactoring. Without getting too technical, lets look at the actual change.
```

  totals = BuyOrderTotals.new(order)
  order.update!(
   # set totals
  )
  order_processor.hold!
  raise Errors::InsufficientInventoryError if order_processor.failed_inventory?
  # in case of failed transaction, we need to rollback this block,
  # but still need to add transaction, so we raise an ActiveRecord::Rollback
  raise ActiveRecord::Rollback if order_processor.failed_payment? || order_processor.requires_action?

  order.update!(
    # set payment
  )
  order.transactions << order_processor.transaction
  PostTransactionNotificationJob.perform_later(order_processor.transaction.id, user_id)
  raise Errors::FailedTransactionError.new(:charge_authorization_failed, order_processor.transaction) if order_processor.failed_payment?
  if order_processor.requires_action?
    # because of an issue with `ActiveRecord::Rollback` we have to force a reload here
    # rollback does not clean the model and calling update on it will raise error
    order.reload.update!(external_charge_id: order_processor.transaction.external_id)
    Exchange.dogstatsd.increment '******'
    raise Errors::PaymentRequiresActionError, order_processor.action_data
  end
end
```

> In original solution we wrapped all of our changes in a database transaction within `order.submit!` which would have done a lock on that record. This was all good since we would ensure data integrity provided by transaction where we want to make sure updates to `order` and `line_items` happen only in case of success. A failure in this block would rollback all changes which is good 👍

Reviewer 3: ```suggestion
In the original solution, we wrapped all of our changes in a database transaction within `order.submit!` to have done a lock on that record. This was all good since we would ensure data integrity provided by, transaction where we want to make sure updates to `order` and `line_items` happen only in case of success. A failure in this block would rollback all changes which is good 👍
```

> But things got complicated once we started have situation that some of the changes during the block should have been preserved even in case of rollback. Specifically we want to make sure a `transaction` is stored on the `order` if it failed payment or requires action.

Reviewer 3: ```suggestion
But things got complicated once some of the changes in the block _should_ have been preserved, even in case of rollback. Specifically we want to make sure a `transaction` is stored on the `order` if it payment fails or requires action.
```

Next question is how to get this to production. We tried to isolate this specific refactoring by:

* Open a PR that only focuses on our refactoring
* Make sure in the PR above we don't touch any API level tests and make sure all these tests still pass. This would give us more confidence that we are not impacting our existing clients.
* Review and merge refactoring PR and test on staging.
* Deploy everything in current pipeline to isolate the refactor deploy.
* Deploy the refactoring PR to production.

## How Did It Go?

This plan worked for us, for the most part. We ended up having to rollback the deploy since we found a bug in a non-API part of our app. From this we learned that even if we already have tests written in different layers of our app, we still need to verify them and make sure they cover all cases. Relying too much on existing tests can often lead to trouble -- verify!

## Our learnings
- Don't be afraid of refactors. They are natural and a healthy engineering tool / practice.
- Ensure that refactor PR's only include refactor-related changes. It's often tempting to fix other things along the way, but those fixes can take place in follow-up PRs.

> - Don't only rely on existing tests. Refactoring is great opportunity to review and verify your tests. Verify them and make sure they cover all scenarios.

Reviewer 3: ```suggestion
- Don't only rely only on existing tests. Refactoring is great opportunity to review and verify your tests. Verify them and make sure they cover all scenarios.
```

> Curious about the PR? At Artsy we believe in Open Source By Default, so check out the code [here](https://github.com/artsy/exchange/pull/475/files).

Reviewer 3: This is awesome! We should add a link to the Open Source by Default section of our principles doc: https://github.com/artsy/README/blob/master/culture/engineering-principles.md#open-source-by-default

Also, what do you think about adding a link to Exchange's repo near the beginning of the post? Just to make it easy for someone skimming to find?

Author: good point, where would you add the link to exchange, in some ways i didn't want to distract people with repo :)

Reviewer 3: Somewhere near the top, since that's where people will probably go to look for it. You can add it after the quote, maybe, in a paragraph surrounded by parens to denote its unimportance to the narrative of the post:

```markdown
> (If you're looking for the code that this blog post refers to, [the Exchange repo is on GitHub](https://github.com/artsy/exchange).)
```
